server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -3286,19 +3321,101 @@ private void doCheckIndex() throws IOException {
     }
 
     Engine getEngine() {
-        Engine engine = getEngineOrNull();
-        if (engine == null) {
-            throw new AlreadyClosedException("engine is closed");
+        engineLock.readLock().lock();
+        try {
+            return getCurrentEngine(false);
+        } finally {
+            engineLock.readLock().unlock();
         }
-        return engine;
     }
 
     /**
      * NOTE: returns null if engine is not yet started (e.g. recovery phase 1, copying over index files, is still running), or if engine is
      * closed.
      */
     public Engine getEngineOrNull() {
-        return this.currentEngineReference.get();
+        engineLock.readLock().lock();
+        try {
+            return getCurrentEngine(true);
+        } finally {
+            engineLock.readLock().unlock();
+        }
+    }
+
+    public <R> R withMutableEngine(Function<Engine, R> operation) {
+        return withEngine(operation, true, false);
+    }
+
+    public <R> R withMutableEngineOrNull(Function<Engine, R> operation) {
+        return withEngine(operation, true, true);
+    }
+
+    public <R> R withImmutableEngine(Function<Engine, R> operation) {
+        return withEngine(operation, false, false);
+    }
+
+    /**
+     * Executes an operation while preventing the shard's engine instance to be changed or closed during the execution. The parameter
+     * {@code requiredMutability} can be used to force the engine to be reset to a given mutable or immutable state before executing the
+     * operation. The parameter {@code allowNoEngine} is used to allow the operation to be executed with a null engine instance, in which
+     * case the {@code requiredMutability} is ignored. When {@code allowNoEngine} is set to {@code `false`} the method will throw an
```

Comment:
Reading the code, if allowNoEngine = true, it still tries to reset the engine according to the requiredMutability. I think you should skip resetting the engine if allowNoEngine = true.

In general, I was a bit baffled with the parameters, but I now see what you're doing with them. Maybe a better naming could be:

```
private <R> R withEngine(Function<Engine, R> operation, boolean requiredMutability, boolean mutable)
```

and if requiredMutability == false, you call getCurrentEngine with true.